### DISCUSSION TITLE
Access to public / admin interfaces from frontend / backend

### DISCUSSION THREAD

**Discussion by alex-pobeditel-2004** [ory/hydra#2697] [Q&A] [answered]
I'm trying to use Hydra as OIDC provider and got confused with Public / Admin interfaces availability from authentication service frontend / backend.

My authentication service has:  
1. Python backend (FastAPI) that implements all the authentication and user management logic
2. JS frontend (Angular) that will provide login / consent forms

I'm following [official instruction on implementing login page](https://www.ory.sh/hydra/docs/guides/login/) and have a big question here:  
Login and consent are accepted/rejected via calls to Hydra\`s **admin interface**. `get_login_request` method also uses admin interface.  

So which part of authentication service must execute this "admin" requests really?

As I understand, if we want to send this requests from JS frontend then we must open some admin endpoints (`/oauth2/auth/requests/`) to the outer world (because it will be called from user's browser). But it also means that user will be able to "accept" his session manually by just sending request with his `login_challenge` to the Hydra\`s admin interface.  

Other way to do it as I see is to chatter with admin interface only from backend. But:

1. Does it mean that each time user is redirected to login page he will need to wait for the next interactions?
\* User on SomeSite clicks "Sign up with MySite" \*  -> SomeSite creates request with valid parameters and starts authentication flow with hydra ->  
`https://hydra_public:4444/oauth2/auth?many=different_parameters` -> Hydra redirects user to MySyte\`s frontend login page ->  
`https://my_frontend:80/login?login_challenge=abcd` -> Frontend sends information to backend to check if we need to show login form ->
`https://my_backend:8080/check-login-request?login_challenge=abcd` -> Backend asks Hydra Admin if user is already authenticated to get things like skip=true ->  
`https://hydra_admin:4445/oauth2/auth/requests/login?login_challenge=abcd` ->  
\* After that (user still have no visible response on his login request) backend tells to the frontend if it must show login form or can just redirect user to callback `redirectTo` \*  
Finally user can see `my_frontend`\`s login page or he's redirected to the callback page  
Am I right here or completely missing the point?

2. Does it mean that when user provided correct login/password backend must return to the frontend **not user\`s access token but just a callback value** (`redirectTo`)?

Hope you can clarify these moments :)

**Comment by vinckr** [comment, marked as answer]
` we must open some admin endpoints (/oauth2/auth/requests/) to the outer world (because it will be called from user's browser). `
The admin endpoint should be available only in the private network!

Ory Hydra by itself does not feature any protection for admin or other endpoints.
In production, you should deploy it behind a reverse proxy/VPN where only your Identity Provider can access it.

See also this document for more information: 
https://www.ory.sh/hydra/docs/next/production/#ory-hydra-behind-an-api-gateway 

Generally, we encourage using [Ory Oathkeeper](https://www.ory.sh/oathkeeper/docs/) for this role, but there are many other options available.

**Reply by vinckr** [reply]
Hey @alex-pobeditel-2004 ,
Just checking in if your question was answered or if there was still something unclear.
I will mark this as answered for now, but feel free to come back to the discussion any time.

Thanks a ton! 🐝

**Reply by alex-pobeditel-2004** [reply]
Hello!  
Finally it seems to me that second way I described in the question was a right way to do it :) At least it still works as expected 😀  
So yes, user should wait while frontend chatters with backend which it its turn talks to ORY Hydra. Good news that this wait will be short if backend, frontend and Hydra are placed near enough 😎
